keel-igor/src/main/kotlin/com/netflix/spinnaker/keel/igor/BuildLifecycleMonitor.kt

```diff
@@ -0,0 +1,197 @@
+      text = "Build succeeded for version ${task.triggeringEvent.artifactVersion}",
+      timestamp = buildMetadata.getCompletedAtInstant()
+    ))
+  }
+
+  private fun publishFailedEvent(task: MonitoredTask, buildMetadata: BuildMetadata) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = FAILED,
+      link = chooseLink(task),
+      text = "Build failed for version ${task.triggeringEvent.artifactVersion}",
+      timestamp = buildMetadata.getCompletedAtInstant()
+    ))
+  }
+
+  private fun publishAbortedEvent(task: MonitoredTask, buildMetadata: BuildMetadata) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = ABORTED,
+      link = chooseLink(task),
+      text = "Build aborted for version ${task.triggeringEvent.artifactVersion}",
+      timestamp = buildMetadata.getCompletedAtInstant()
+    ))
+  }
+
+  private fun publishUnknownStatusEvent(task: MonitoredTask, status: String?) {
+    log.warn("Unknown status $status while monitoring ${task.triggeringEvent}")
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = UNKNOWN,
+      link = chooseLink(task),
+      text = "Build status unknown for version ${task.triggeringEvent.artifactVersion}"
+    ))
+  }
+
+  private fun publishUnknownEvent(task: MonitoredTask) {
+    log.warn("Unable to monitor build for ${task.triggeringEvent} because at least one required data field is missing")
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = UNKNOWN,
+      link = chooseLink(task),
+      text = "Build status unknown for version ${task.triggeringEvent.artifactVersion}"
+    ))
+  }
+
+  override fun publishExceptionEvent(task: MonitoredTask) {
+    publisher.publishEvent(task.triggeringEvent.copy(
+      status = UNKNOWN,
+      link = chooseLink(task),
+      text = "Failed to monitor build of version ${task.triggeringEvent.artifactVersion}" +
+        " because we could not get the status ${lifecycleConfig.numFailuresAllowed} times. Status unknown."
+    ))
+  }
+
+  fun chooseLink(task: MonitoredTask): String? {
+    val buildData = parseBuildData(task)
+    val app = runBlocking {
+      try {
+        front50Service.applicationByName(buildData.application)
+      } catch (e: HttpException) {
+        if (e.isNotFound) {
+          null
+        } else {
+          throw e
+        }
+      }
+    }
+
+    if (app == null || app.dataSources?.disabled?.contains("integration") == true) {
+      // app not found or ci explicitly disabled
+      return jenkinsLink(buildData)
+    }
+
+    // ci not explicitly disabled, check and see if ci view is configured
+    return if (app.repoProjectKey != null && app.repoSlug != null && app.repoType != null) {
+      buildUidToLink(task)
+    } else {
+      jenkinsLink(buildData)
+    }
+  }
+
+  private fun buildUidToLink(task: MonitoredTask): String =
+    "$spinnakerBaseUrl/#/applications/${task.triggeringEvent.data["application"]}/builds/${task.link}/logs"
```

Comment:
I'm confused about why we'd want to do this server side

Reply:
Because... erik asked me to. He said we could do it on the UI side, but it would be better to just have the data.

Also, we don't have support right now for an "alternate link". It makes sense to calculate the link shown on the back end because we can do more complex logic more easily.